jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,465 @@
+                    // See correspondent code in resume().
+                    int suspended = AtomicBiInteger.getLo(encoded);
+                    if (!state.compareAndSet(encoded, 0, suspended + 1))
+                        continue;
+                    suspend(request, response, callback);
+                }
+                return true;
+            }
+        }
+    }
+
+    /**
+     * <p>Returns the priority of the given suspended request,
+     * a value greater than or equal to {@code 0}.</p>
+     * <p>Priority {@code 0} is the lowest priority.</p>
+     * <p>The set of returned priorities should be stable over
+     * time, typically constrained in the range {@code 0-10}.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request, a value {@code >= 0}
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    /**
+     * <p>Fails the given suspended request/response with the given error code and failure.</p>
+     * <p>This method is called only for suspended requests, in case of timeout while suspended,
+     * or in case of failure when trying to handle a resumed request.</p>
+     *
+     * @param request the request to fail
+     * @param response the response to fail
+     * @param callback the callback to complete
+     * @param status the failure status code
+     * @param failure the failure
+     */
+    protected void fail(Request request, Response response, Callback callback, int status, Throwable failure)
+    {
+        Response.writeError(request, response, callback, status, null, failure);
+    }
+
+    private boolean forward(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} forwarding {}", this, request);
+        request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
+        return super.handle(request, response, callback);
+    }
+
+    private void suspend(Request request, Response response, Callback callback)
+    {
+        int priority = Math.max(0, getPriority(request));
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} suspending priority={} {}", this, priority, request);
+        Entry entry = new Entry(request, response, callback, priority);
+        queues.compute(priority, (k, v) ->
+        {
+            if (v != null)
+                return v;
+            priorities.add(priority);
+            return new ConcurrentLinkedQueue<>();
+        }).offer(entry);
+        timeouts.schedule(entry);
+    }
+
+    private void resume()
+    {
+        // Only increment the permits if there are no suspended requests.
+        // In this way, new requests will not steal the permits,
+        // so that already suspended requests will not starve.
+        while (true)
+        {
+            for (Integer priority : priorities)
```

Comment:
This logic looks like a scheduler. It may be worth extracting that into a utility class.